group-coordinator/src/main/java/org/apache/kafka/coordinator/group/RecordSerDe.java

```diff
@@ -0,0 +1,161 @@
+import org.apache.kafka.coordinator.group.generated.OffsetCommitKey;
+import org.apache.kafka.coordinator.group.generated.OffsetCommitValue;
+import org.apache.kafka.coordinator.group.runtime.CoordinatorLoader;
+import org.apache.kafka.coordinator.group.runtime.PartitionWriter;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+
+/**
+ * Serializer/Deserializer for {{@link Record}}.
+ */
+public class RecordSerDe implements PartitionWriter.Serializer<Record>, CoordinatorLoader.Deserializer<Record> {
+    @Override
+    public byte[] serializeKey(Record record) {
+        // Record does not accept a null key.
+        return MessageUtil.toVersionPrefixedBytes(
+            record.key().version(),
+            record.key().message()
+        );
+    }
+
+    @Override
+    public byte[] serializeValue(Record record) {
+        // Tombstone is represented with a null value.
+        if (record.value() == null) {
+            return null;
+        } else {
+            return MessageUtil.toVersionPrefixedBytes(
+                record.value().version(),
+                record.value().message()
+            );
+        }
+    }
+
+    @Override
+    public Record deserialize(
+        ByteBuffer keyBuffer,
+        ByteBuffer valueBuffer
+    ) throws RuntimeException {
+        final short keyVersion = readVersion(keyBuffer, "key");
+        final ApiMessage keyMessage = apiMessageKeyFor(keyVersion);
+        readMessage(keyMessage, keyBuffer, keyVersion, "key");
+
+        if (valueBuffer == null) {
+            return new Record(
+                new ApiMessageAndVersion(keyMessage, keyVersion),
+                null
+            );
+        }
+
+        final ApiMessage valueMessage = apiMessageValueFor(keyVersion);
+        final short valueVersion = readVersion(valueBuffer, "value");
+        readMessage(valueMessage, valueBuffer, valueVersion, "value");
+
+        return new Record(
+            new ApiMessageAndVersion(keyMessage, keyVersion),
+            new ApiMessageAndVersion(valueMessage, valueVersion)
+        );
+    }
+
+    private short readVersion(ByteBuffer buffer, String name) throws RuntimeException {
+        try {
+            return buffer.getShort();
+        } catch (BufferUnderflowException ex) {
+            throw new RuntimeException(String.format("Could not read version from %s's buffer.", name));
+        }
+    }
+
+    private void readMessage(ApiMessage message, ByteBuffer buffer, short version, String name) throws RuntimeException {
+        try {
+            message.read(new ByteBufferAccessor(buffer), version);
+        } catch (RuntimeException ex) {
+            throw new RuntimeException(String.format("Could not read record from %s's buffer due to: %s.",
+                name, ex.getMessage()), ex);
+        }
+    }
+
+    private ApiMessage apiMessageKeyFor(short version) {
```

Comment:
should these be 
`short recordType/keyVersion` for both key & value? it would be good to clarify since the versions of keys and values represent different things.